torch_np/_funcs_impl.py

```diff
@@ -909,6 +909,27 @@ def take_along_axis(arr: ArrayLike, indices: ArrayLike, axis):
     return torch.take_along_dim(arr, indices, axis)
 
 
+def put(
+    a: NDArray,
+    ind: ArrayLike,
+    v: ArrayLike,
+    mode: NotImplementedType = "raise",
+):
+    v = v.type(a.dtype)
+    # If ind is larger than v, broadcast v to the would-be resulting shape. Any
+    # unnecessary trailing elements are then trimmed.
+    if ind.numel() > v.numel():
+        result_shape = torch.broadcast_shapes(v.shape, ind.shape)
+        v = torch.broadcast_to(v, result_shape)
```

Comment:
Is this correct? If `ind = torch.arange(4)` and `v = torch.arange(3).unsqueeze(1)`, then we'll broadcast `v`, and we'll then trim it in the next line. The issue here is that by doing the broadcasting in the last dimension, we'll be copying the elements `0, 0, 0, 0`, rather than `0, 1, 2, 0`. For the elements to repeat as we want when doing the `flatten` below we want to do the broadcasting on the left-most dimension. This was also a bug in my proposed implementation. You can do that by using the ceil computation I proposed + `unsqueeze(0)` + `expand`.

Reply:
Ah yeah you're right, for reference

```python
>>> a = np.arange(5)
>>> ind = np.arange(4)
>>> v = np.arange(3).unsqueeze(1)
>>> np.put(a, ind, v)
>>> a
array([0, 1, 2, 0, 4])
```

```python
>>> x = tnp.arange(5)
>>> ind = tnp.arange(4)
>>> v = tnp.asarray([[0], [1], [2]])
>>> tnp.put(x, ind, v)
>>> x
array_w([0, 0, 0, 0, 4])
```

So Ill need to do more exploration/testing which still satisfies the existing problem I came across. Will leave for next week :sweat_smile:

<details>
<summary>FWIW dirty/un-minified failing case I alluded to before when using repeat scenario before</summary>

```python
     AssertionError: 
     Arrays are not equal
     
     Mismatched elements: 1 / 2 (50%)
      x: array_w([[[ True, False]]])
      y: array_w([[[True, True]]])
     Falsifying example: test_put(
         np_x=array([[[False, False]]]),
         data=data(...),
     )
     Draw 1 (ind): (array([[-1],[-1]]), array([[0],[0]]), array([[0],[0]]))
     Draw 2 (v): array([[False],
            [ True]])
     (after put) np_x=array([[[ True,  True]]])
     tnp_x=array_w([[[False, False]]])
     (after put) tnp_x=array_w([[[ True, False]]])


     Mismatched elements: 2 / 12 (16.7%)
      x: array_w([[ True, False, False,  True],
             [False, False, False, False],
             [False, False, False, False]])
      y: array_w([[ True, False,  True, False],
             [False, False, False, False],
             [False, False, False, False]])
     Falsifying example: test_put(
         np_x=array([[False, False, False, False],
                [False, False, False, False],
                [False, False, False, False]]),
         data=data(...),
     )
     Draw 1 (ind): (array([[1],
             [2]]), array([[3],
             [0]]))
     Draw 2 (v): array([[False],
            [ True]])
     (after put) np_x=array([[ True, False,  True, False],
            [False, False, False, False],
            [False, False, False, False]])
     tnp_x=array_w([[False, False, False, False],
             [False, False, False, False],
             [False, False, False, False]])
     (after put) tnp_x=array_w([[ True, False, False,  True],
             [False, False, False, False],
             [False, False, False, False]])
```

</details>

Reply:
Pushed a fix using `repeat` again, and make sure its tested for in `test_put`.

Notably repeating just on the 0th axis seems the right call for our `put()` to align with NumPy-proper, as other ways of constructing `*sizes` falls short (i.e. Hypothesis will come up with failing examples in `test_put`).